File: src/LeetCode/Algorithms/Java/Path_Sum_III.java
package LeetCode.Algorithms.Java;

import java.util.LinkedList;
import java.util.List;

// https://leetcode.com/problems/path-sum-iii/
public class Path_Sum_III {
	public static class TreeNode {
		int val;
		TreeNode left;
		TreeNode right;
		
		TreeNode(int x) {
			this.val = x;
		}
	}
	
	public static void main(String[] args) {
		int sum = 8;
		TreeNode root = new TreeNode(10);
		
		root.left = new TreeNode(5);
		root.right = new TreeNode(-3);
		root.left.left = new TreeNode(3);
		root.left.right = new TreeNode(2);
		root.left.left.left = new TreeNode(3);
		root.left.left.right = new TreeNode(-2);
		root.left.right = new TreeNode(2);
		root.left.right.right = new TreeNode(1);
		root.right.right = new TreeNode(11);
		
		System.out.println(pathSum(root, sum));
	}
	
	public static int pathSum(TreeNode root, int sum) {
		if (root == null) {
			return 0;
		}
		
		return recurrentPathSum(root, sum) + pathSum(root.left, sum) + pathSum(root.right, sum);
	}
	
	public static int recurrentPathSum(TreeNode root, int sum) {
		if (root == null) {
			return 0;
		}
		
		return ((root.val == sum) ? 1 : 0) + recurrentPathSum(root.left, sum - root.val) + recurrentPathSum(root.right, sum - root.val);
	}
}

File: src/LeetCode/Algorithms/JavaScript/Path_Sum_III.js
/**
 * https://leetcode.com/problems/path-sum-iii/
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} sum
 * @return {number}
 */
var pathSum = function(root, sum) {
	if (!root) {
		return 0;
	}
	
	const recurrentPathSum = (root, sum) => {
		if (!root) {
			return 0;
		}
		
		return ((root.val === sum) ? 1 : 0) + recurrentPathSum(root.left, sum - root.val) + recurrentPathSum(root.right, sum - root.val);
	};
	
	return recurrentPathSum(root, sum) + pathSum(root.left, sum) + pathSum(root.right, sum);
};

function TreeNode(val) {
	this.val = val;
	this.left = this.right = null;
}

const sum = 8;
let root = new TreeNode(10);

root.left = new TreeNode(5);
root.right = new TreeNode(-3);
root.left.left = new TreeNode(3);
root.left.right = new TreeNode(2);
root.left.left.left = new TreeNode(3);
root.left.left.right = new TreeNode(-2);
root.left.right = new TreeNode(2);
root.left.right.right = new TreeNode(1);
root.right.right = new TreeNode(11);

console.log(pathSum(root, sum));